### DISCUSSION TITLE
How do you pause the physics simulation?

### DISCUSSION THREAD

**Discussion by Pyroglyph** [idanarye/bevy-tnua#21] [Q&A] [answered]
I'd like to be able to just pause the simulation both for using with a pause menu and for implementing mechanics like a dash ability that sends you shooting linearly in a single direction for a short time while removing control from the player ([a la Celeste](https://youtu.be/yorTG9at90g?t=421)).
There doesn't seem to be any obvious way of doing this other than removing a core physics component (like `TnuaRapier2dIOBundle`), but that doesn't feel right to do. There has to be a better way, right?

**Comment by idanarye** [comment, marked as answer]
For the pause menu, you can use [`TnuaSystemSet`](https://docs.rs/bevy-tnua/0.6.1/bevy_tnua/struct.TnuaSystemSet.html). Bevy allows you to put conditions on existing system sets, so you can just put your own condition on it. This is how it is done in the demos:

https://github.com/idanarye/bevy-tnua/blob/35780719300907075b4e2e671b46249514687890/examples/platformer_3d.rs#L40

Keep in mind that you also need to [pause Rapier](https://docs.rs/bevy_rapier3d/latest/bevy_rapier3d/plugin/struct.RapierConfiguration.html#structfield.physics_pipeline_active) while in the pause menu.

As for removing controls from the player - I'm afraid it is not currently supported. I do want to implement things like dashes as part of Tnua (#6). For now, you don't have to remove the entire bundle - removing the [`TnuaMotor`](https://docs.rs/bevy-tnua/0.6.1/bevy_tnua/struct.TnuaMotor.html) should be enough.

**Reply by Pyroglyph** [reply]
Thanks, that works great! I didn't know you could get resources in `run_if`, which is useful to know!

By the way, when I said "removing control from the player [while dashing]" I didn't mean "remove the `TnuaPlatformerControls` component", I meant "ignore player inputs" which was easy enough to achieve.

One more thing. I set up the condition to not run while dashing, and in the dash system I directly modify the `Velocity` component. It seems to work how I want it to (dash linearly for X time and then maintain velocity when physics is switched back on), but I'd like to know if there's some edge case that might break something. Is there a better way of doing it or is this fine?

**Reply by idanarye** [reply]
> Thanks, that works great! I didn't know you could get resources in `run_if`, which is useful to know!

The function you pass to `run_if` is a Bevy system that returns a boolean. It can access anything a system does - components and resources alike.

> but I'd like to know if there's some edge case that might break something. Is there a better way of doing it or is this fine?

Now that you mention it - yes. Tnua uses Rapier's `ExternalForce` to play with the gravity the character experiences. If you just disable the systems, and there was already something in that component, Tnua's `apply_motors_system` will not override it - and Rapier will keep applying that force. You may want to reset it manually.

**Comment by idanarye** [comment]
Starting from version 0.7:

1. The correct way to pause the simulation entirely is to set `RapierConfiguration.physics_pipeline_active` to `false`. Unlike previous versions,, starting from version 0.7 doing this will automatically pause the `TnuaSystemSet` systems, so no additional interference is required.
2. The correct way to temporarily disable Tnua for a specific entity is to add the [`TnuaToggle`](https://docs.rs/bevy-tnua/0.7.0/bevy_tnua/enum.TnuaToggle.html) component to it. Change it to `Disabled` or `SenseOnly` when you need to control the rigid body without Tnua, and change it back to `Enabled` when you want Tnua to take over again.
